### PR TITLE
Add cupsArrayDup test and overflow protection in array allocation

### PR DESCRIPTION
**Pull request by tanjiroK-coder**
I fixed a bug in `cupsArrayDup()` where the code could calculate a huge allocation size and accidentally overflow the multiplication when doing `num_elements * sizeof(void *)`
So now the patch checks that size safely before calling `malloc()` and also uses the same safe check for hash allocations.

I also added a regression test in `tests/test_array_dup.c` to make sure the function rejects an overflowed size instead of trying to allocate it.

This is a real hardening fix for a memory-safety issue, not just a small cleanup.

### CODE REVIEW

File: cups/array.c [hunk 1]
```diff
@@ -58,6 +58,18 @@ struct _cups_array_s			// CUPS array structure
 static bool	cups_array_add(cups_array_t *a, void *e, bool insert);
 static size_t	cups_array_find(cups_array_t *a, void *e, size_t prev, int *rdiff);
 
+static bool	cups_array_multiply(size_t count, size_t size, size_t *bytes);
```
Comment (michaelrsweet) [inline cups/array.c:62]: No, let's just define a reasonable maximum for `num_elements` (maybe 2^30-1 which is ~1 billion elements) and limit it accordingly.

```c
#define _CUPS_ARRAY_MAX_ELEM 0x1fffffff
```

File: cups/array.c [hunk 2]
```diff
@@ -280,8 +292,10 @@ cupsArrayDup(cups_array_t *a)		// I - Array
   if (a->num_elements)
   {
     // Allocate memory for the elements...
-    da->elements = malloc((size_t)a->num_elements * sizeof(void *));
-    if (!da->elements)
+    size_t bytes;
```
Comment (michaelrsweet) [inline cups/array.c:299]: Assuming that you are duplicating an array you've created using the normal APIs, it is unlikely that duplicating will overflow. But for completeness-sake we can verify that `a->num_elements <= _CUPS_MAXELEM` before trying to allocate that much memory.

File: cups/array.c [hunk 3]
```diff
@@ -306,6 +320,23 @@ cupsArrayDup(cups_array_t *a)		// I - Array
     da->alloc_elements = a->num_elements;
   }
 
+  /* Copy hash table if present so duplicated arrays preserve hashed lookups */
```
Comment (michaelrsweet) [inline cups/array.c:327]: Copying the hash is important, but again the size should already be limited.

File: cups/array.c [hunk 4]
```diff
@@ -653,17 +684,18 @@ cupsArrayNew(cups_array_cb_t  f,	// I - Comparison callback function or `NULL` f
 
   if (hsize > 0 && hf)
   {
+    size_t bytes;
```
Comment (michaelrsweet) [inline cups/array.c:691]: Add a `_CUPS_MAXHASH` limit and ignore the hashing if it exceeds the limit, e.g.:

```c
#define _CUPS_MAXHASH 0x1fffffff

...

if (hsize > 0 && hsize <= _CUPS_MAXHASH && hf)
{
  ...
}
```

File: cups/array.c [hunk 5]
```diff
@@ -863,7 +895,10 @@ cups_array_add(cups_array_t *a,		// I - Array
     else
       count = a->alloc_elements + 1024;
 
-    if ((temp = realloc(a->elements, count * sizeof(void *))) == NULL)
+    size_t bytes;
```
Comment (michaelrsweet) [inline cups/array.c:902]: Limit count to `_CUPS_MAXELEMS`.

File: tests/README.md [hunk 6]
```diff
@@ -0,0 +1,19 @@
+Test helper: cupsArrayDup
```
Comment (michaelrsweet) [inline tests/README.md:1]: NO.

The "testarray.c" unit test program tests the `cupsArray` API. Add to that please.

File: tests/test_array_dup.c [hunk 7]
```diff
@@ -0,0 +1,100 @@
+#include <stdio.h>
```
Comment (michaelrsweet) [inline tests/test_array_dup.c:1]: This should be part of the existing "testarray.c".

**Review by michaelrsweet — CHANGES_REQUESTED** [review]
I appreciate the attention to detail, although it is quite likely that your process would be killed (or get a failed malloc/realloc) long before the overflow is reached.

My preference is to use a reasonable limit on the number of elements and hash size (2^30-1) that will work for 32-bit and 64-bit platforms. And any new tests need to be added to the existing "testarray.c" unit test program.
